src/coreclr/jit/importercalls.cpp

```diff
@@ -1929,111 +2047,203 @@ void Compiler::impPopArgsForUnmanagedCall(GenTreeCall*        call,
+
+    if (swiftErrorIndex != sig->numArgs)
     {
-        GenTree* thisPtr = call->gtArgs.GetArgByIndex(0)->GetNode();
-        impBashVarAddrsToI(thisPtr);
-        assert(thisPtr->TypeGet() == TYP_I_IMPL || thisPtr->TypeGet() == TYP_BYREF);
+        *swiftErrorArg = call->gtArgs.GetArgByIndex(swiftErrorIndex);
     }
 
-    unsigned short argIndex = 0;
-    for (CallArg& arg : call->gtArgs.Args())
+    // Now expand struct args that must be lowered into primitives
+    unsigned argIndex = 0;
+    for (CallArg* arg = call->gtArgs.Args().begin().GetArg(); arg != nullptr; argIndex++)
     {
-        GenTree* argNode = arg.GetEarlyNode();
+        if (!varTypeIsStruct(arg->GetSignatureType()))
+        {
+            arg = arg->GetNext();
+            continue;
+        }
 
-        // We should not be passing gc typed args to an unmanaged call.
-        if (varTypeIsGC(argNode->TypeGet()))
+        if (varTypeIsSIMD(arg->GetSignatureType()))
         {
-            // Tolerate byrefs by retyping to native int.
-            //
-            // This is needed or we'll generate inconsistent GC info
-            // for this arg at the call site (gc info says byref,
-            // pinvoke sig says native int).
-            //
-            if (argNode->TypeGet() == TYP_BYREF)
-            {
-                argNode->ChangeType(TYP_I_IMPL);
-            }
-            else
-            {
-                assert(!"*** invalid IL: gc ref passed to unmanaged call");
-            }
+            IMPL_LIMITATION("SIMD types are currently unsupported in Swift calls");
         }
 
-#ifdef SWIFT_SUPPORT
-        if (argIndex == swiftErrorIndex)
+        JITDUMP("  Argument %u is a struct [%06u]\n", argIndex, dspTreeID(arg->GetNode()));
+
+        assert(arg->GetNode()->OperIsLocalRead());
+        GenTreeLclVarCommon* structVal = arg->GetNode()->AsLclVarCommon();
+
+        CallArg* insertAfter = arg;
+        // For the self arg, change it from the SwiftSelf struct to a
+        // TYP_I_IMPL primitive directly. It must also be marked as a well
+        // known arg because it has a non-standard calling convention.
+        if (argIndex == swiftSelfIndex)
         {
-            // Found the SwiftError* arg
-            assert(swiftErrorArg != nullptr);
-            *swiftErrorArg = &arg;
+            assert(arg->GetNode()->OperIsLocalRead());
+            GenTree*   primitiveSelf = gtNewLclFldNode(structVal->GetLclNum(), TYP_I_IMPL, structVal->GetLclOffs());
+            NewCallArg newArg = NewCallArg::Primitive(primitiveSelf, TYP_I_IMPL).WellKnown(WellKnownArg::SwiftSelf);
+            insertAfter       = call->gtArgs.InsertAfter(this, insertAfter, newArg);
         }
-        else if (argIndex == swiftSelfIndex)
+        else
         {
-            // Found the SwiftSelf arg
-            arg.SetWellKnownArg(WellKnownArg::SwiftSelf);
+            CORINFO_SWIFT_LOWERING* lowering = lowerings[argIndex];
+
+            if (lowering == nullptr)
```

Comment:
`(lowering == nullptr)` indicates that either the arg should be passed by reference, or that we never called `getSwiftLowering` for this arg (because it's a `SwiftSelf`, or it's not a value type), as the `lowerings` array is zero-initialized, right? It would sacrifice some perf, but this might read better if we always save the result of `getSwiftLowering` and check `lowering->byReference` here instead; that way, we know `getSwiftLowering` was never called for this arg if `lowerings[argIndex] == nullptr`, and can assert on it.

I don't have a strong preference on this, though.

Reply:
The `SwiftSelf` case is handled before we get here, but other than that you're right, `lowering == nullptr` here means that the arg is passed by reference. I'll switch it to always save the lowerings for all the struct args and add an assert -- I agree that seems better.